docs: fill out react native and ios docs topics (#104)
spokestack/spokestack-website: content/docs/React Native/nlu.md

---
title: NLU on React Native

> navId: NLU (React Native)

Reviewer: File needs to be added to `nav.json`

description: Understanding the React Native NLU API
draft: false
---

> This is a companion to the [NLU concept guide](docs/Concepts/nlu), which discusses the NLU subsystem holistically. Here we'll talk about usage issues specific to the React Native client library.

Reviewer 2: ```suggestion
This is a companion to the [NLU concept guide](/docs/Concepts/nlu), which discusses the NLU subsystem holistically. Here we'll talk about usage issues specific to the React Native client library.
```

## Configuration

As mentioned in the [Getting Started](getting-started) guide, initializing the Spokestack NLU is done using the same interface as the other Spokestack components:

```javascript
Spokestack.initialize({
  // omitting speech pipelineline and tts configuration
  input: '',
  stages: [],
  tts: {},
  properties: {},
  nlu: {

>     // NLU settings. Only set these if you are calling `Spokestack.classify`.

Reviewer 2: ```suggestion
```
This guide is all about NLU, so assume these are required

    'nlu-model-path': YOUR_NLU_MODEL_PATH, // string filesystem path to nlu model
    'nlu-metadata-path': YOUR_NLU_METADATA_PATH, // string filesystem path to nlu metadata
    'wordpiece-vocab-path': YOUR_NLU_VOCABULARY_PATH // string filesystem path to nlu vocab
  }
})
```

> The configuration properties above refer to the three required files for the Spokestack NLU model you're using.

Reviewer 2: Is it worth having a quick explainer here about how the filesystem paths work, maybe an example using rn-fetch's API? People seem to get tripped up here.

Author: oh yeah i actually meant to do this here, thanks

## Usage

> When it comes time to classify an utterance, Spokestack's NLU does all the heavy lifting in the background thread and returns the classification result via an event body.

Reviewer 2: ```suggestion
When it comes time to classify an utterance, Spokestack's NLU does all the heavy lifting on a background thread and returns the classification result via an event body.
```

First register the app to receive the `onClassification` event:

```javascript
Spokestack.onClassification = (e) => {
  console.log(e.result.intent)
}
```

> Then just call `classify` with the text you want to classify. The second argument, a dictionary, is currently unused.

Reviewer 2: ```suggestion
Then just call `classify` with the text you want to classify. The second argument, a plain object, is currently unused.
```
The RN module could even hide it from its public API (or add a convenience method that omits it) until we have a service that uses it.

```javascript
Spokestack.classify('Tea, earl grey, hot', {})
```

> Your `onClassification` event will receive a dictionary with the following structure:

Reviewer 2: ```suggestion
Your `onClassification` callback will receive an object with the following structure:
```

```javascript
{
  "event":"classification"

>   ,"result": {

Reviewer 2: I get that this is technically valid/easier to update by hand, but do you ever actually _see_ JSON in the wild like this?

    "confidence":"0.9999006"
    ,"intent":"RecipeIntent"
    ,"slots": {
      "Item":{

>         "rawValue":"tea earl grey hot"

Reviewer 2: Looks like we lost the capitalization/punctuation from the original utterance here.

        ,"value":"tea"
        ,"type":"entity"
      }
    }
  }
  ,"error":""
}
```

> From there, your app takes over and acts on the intent and slots that have been classified!

Reviewer 2: ```suggestion
From there, your app takes over and acts on the intent and slots!
```
